src/test/server.ts

```diff
@@ -0,0 +1,36 @@
+import { exec } from 'child_process';
+import net from 'net';
+import { monotonicTime, raceAgainstDeadline } from './util';
+
+export async function launchWebServer({command, port, timeout = 10000}: {command: string, port: number, timeout?: number}) {
+    const webServer = exec(command);
+    const cancellationToken = {canceled: false};
+    const {timedOut} = await raceAgainstDeadline(waitForSocket(port, 100, cancellationToken), timeout + monotonicTime());
+    cancellationToken.canceled = true;
+    if (timedOut) {
+        webServer.kill();
+        throw new Error(`failed to start web server on port ${port} via "${command}"`);
+    }
+    process.env.PW_BASE_URL = `http://localhost:${port}`;
+    return webServer;
+}
+async function waitForSocket(port: number, delay: number, cancellationToken: {canceled: boolean}) {    
+    while (!cancellationToken.canceled) {
+        const connected = await new Promise((resolve) => {
+        const conn = net
+            .connect(port)
```

Comment:
It would be cool if we could verify that at least the server responds. Its often that a server is started on the TCP level, then express e.g. will use it and don't respond until the actual dev server is started. Thats takes sometimes 30-120 seconds to boot up then working server.

Besides that we could think about checking the status codes.

Reply:
Theres a grace period in the browser when you `goto` an unresponsive browser. For Chromium it seems to be about 500ms if nothing is listening on the port, and then it's more like 30 seconds if something is listening. So I think that waiting for something to be listening is the most important part.

If the server takes 120 seconds to start up that is still going to break. `jest-dev-server` has a default timeout of 5 seconds. What's going on that causes a server to take 2 minutes?